File: algorithms/comparison/selection_sort.py
def selection_sort(arr: list[int]) -> list[int]:
    """Selection sort

    The algorithm sorts a list by comparing one item with the remaining items in the list
    and swap that item with the smallest one. Continue with the second item until the last one.

    Args:
        arr (list): An unsorted list of integers

    Returns:
        arr (list): A sorted array in increasing order
    """
    if not arr:
        return []

    # Make a copy to avoid modifying the original
    result = arr.copy()

    for i in range(len(result)):
        min_idx = i
        for j in range(i + 1, len(result)):
            if result[j] < result[min_idx]:
                min_idx = j

        result[i], result[min_idx] = result[min_idx], result[i]

    return result

File: algorithms/comparison/cpp_version/selection_sort.cpp
#include <iostream>
#include <vector>

using namespace std;

std::vector<int> selection_sort(std::vector<int> arr)
{
    int n = arr.size();
    for (int i = 0; i < n; i++)
    {
        for (int j = i; j < n; j++)
        {
            if (arr[i] > arr[j])
            {
                std::swap(arr[i], arr[j]);
            };
        };
    };
    return arr;
}

int main()
{
    std::vector<int> arr = {10, 9, 8, 7, 6, 5, 4, 3, 2, 1};
    std::vector<int> sorted_arr = selection_sort(arr);

    for (int num : sorted_arr)
    {
        cout << num << endl;
    };

    return 0;
}